ios/src/TGFontConverter.mm

```diff
@@ -0,0 +1,154 @@
+//
+//  TGFontConverter.m
+//  tangram
+//
+//  Created by Karim Naaji on 11/01/16.
+//
+//
+
+#import "TGFontConverter.h"
+
+#include <cstdio>
+
+struct FontHeader {
+    int32_t fVersion;
+    uint16_t fNumTables;
+    uint16_t fSearchRange;
+    uint16_t fEntrySelector;
+    uint16_t fRangeShift;
+};
+
+typedef struct FontHeader FontHeader;
+
+struct TableEntry {
+    uint32_t fTag;
+    uint32_t fCheckSum;
+    uint32_t fOffset;
+    uint32_t fLength;
+};
+
+typedef struct TableEntry TableEntry;
+
+static uint32_t CalcTableCheckSum(const uint32_t* table, uint32_t numberOfBytesInTable)
+{
+    uint32_t sum = 0;
+    uint32_t nLongs = (numberOfBytesInTable + 3) / 4;
+    while (nLongs-- > 0) {
+       sum += CFSwapInt32HostToBig(*table++);
+    }
+    return sum;
+}
+
+static uint32_t CalcTableDataRefCheckSum(CFDataRef dataRef)
+{
+    const uint32_t* dataBuff = (const uint32_t *)CFDataGetBytePtr(dataRef);
+    uint32_t dataLength = (uint32_t)CFDataGetLength(dataRef);
+    return CalcTableCheckSum(dataBuff, dataLength);
+}
+
+@implementation TGFontConverter
+
+//Reference:
+//http://skia.googlecode.com/svn-history/r1473/trunk/src/ports/SkFontHost_mac_coretext.cpp
+//https://gist.github.com/Jyczeal/1892760
+
++ (unsigned char *)fontDataForCGFont:(CGFontRef)cgFont size:(size_t *)size
+{
+    if (!cgFont) {
+        return nil;
+    }
+
+    CFRetain(cgFont);
+
+    CFArrayRef tags = CGFontCopyTableTags(cgFont);
+    int tableCount = CFArrayGetCount(tags);
+
+    size_t* tableSizes = (size_t*)malloc(sizeof(size_t) * tableCount);
+    memset(tableSizes, 0, sizeof(size_t) * tableCount);
+
+    BOOL containsCFFTable = NO;
+
+    size_t totalSize = sizeof(FontHeader) + sizeof(TableEntry) * tableCount;
+
+    for (int index = 0; index < tableCount; ++index) {
+
+        //get size
+        size_t tableSize = 0;
+        intptr_t aTag = (intptr_t)CFArrayGetValueAtIndex(tags, index);
+
+        if (aTag == 'CFF ' && !containsCFFTable) {
+            containsCFFTable = YES;
+        }
+
+        CFDataRef tableDataRef = CGFontCopyTableForTag(cgFont, aTag);
+        if (tableDataRef != NULL) {
+            tableSize = CFDataGetLength(tableDataRef);
+            CFRelease(tableDataRef);
+        }
+        totalSize += (tableSize + 3) & ~3;
+
+        tableSizes[index] = tableSize;
+    }
+
+    unsigned char* stream = (unsigned char*)malloc(totalSize);
+
+    memset(stream, 0, totalSize);
+    char* dataStart = (char*)stream;
+    char* dataPtr = dataStart;
+
+    // compute font header entries
+    uint16_t entrySelector = 0;
+    uint16_t searchRange = 1;
+
+    while (searchRange < tableCount >> 1) {
```

Comment:
This part gets the closest power of two after tableCount and gets the 4th next power of two after it.

This is described here in the _Organization of an OpenType Font_ https://www.microsoft.com/typography/otspec/otff.htm.

It is described as "(Maximum power of 2 <= numTables) x 16."
So you can see the x16 as `searchRange <<= 4;` and the `(Maximum power of 2 <= numTables)` as the while loop.